apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/AbstractSpan.java

```diff
@@ -107,6 +109,11 @@
 
     protected final AtomicReference<Span> bufferedSpan = new AtomicReference<>();
 
+    @Nullable
+    private OTelSpanKind otelKind = null;
+
+    private final Map<String, Object> otelAttributes = new HashMap<>();
+
```

Comment:
I will try to look at an alternative for this, maybe close to what you suggested.

Reply:
That's right, we shouldn't add a state to `OTelSpan`, the `AbstractSpan` would still own the bridge context. It's not where the data resides, only where the implementation resides. I prefer the bridge to maintain this part of the context, which is tightly coupled with the OTel API.

This is what I propose (calling it `FreeFormContext` because the internal context already has room for custom context):
We add another context type that has a very loose API:
```
interface FreeFormContext extends Recyclable
{
  serialize(JsonWriter);
}
```
Any plugin can implement it and be responsible of proper serialization of it.
Then we add another context property to spans:
```
public abstract class AbstractSpan<T extends AbstractSpan<T>> implements Recyclable {
    ....
    public abstract AbstractContext getContext();
    ....
    public abstract void setFreeFormContext(FreeFormContext);
    public abstract FreeFormContext getFreeFormContext();
}
```
The bridge will have a `OTelSpanContext implements FreeFormContext` and instead of doing:
```
span.getOtelAttributes().put("key", "value");
```
`OTelSpan` would do something like:
```
((OTelSpanContext)span.getFreeFormContext()).setAttribute("key", "value");
```
I hope this makes sense.

Reply:
I see what you mean here, but I'm not sure that we can generically handle the JSON serialization part in an elegant way.
We could even allow the plugin to create dedicated sub-classes of `Transaction` and `Span` for that purpose.

For example, we have to assume that the serialized form will always be in the same place in JSON.
In this case, we have an extra `otel` attribute: `span.otel` object with `span.otel.kind` and `span.otel.attributes`, but any other implementation could use a different structure.

While I like the idea to keep the core of the agent as OTel-agnostic as possible, I don't think that we should handle this right now:
- we only have a single implementation, thus it's probably too early to create an abstraction that will fit a future abstraction.
- I don't see any other possible implementation candidate for this.
- it's only a couple of fields, thus it seems an acceptable compromise, and we can always revisit this later.

Reply:
> but I'm not sure that we can generically handle the JSON serialization part in an elegant way.

Wouldn't a copy+paste of the current serialization work without any change?

> For example, we have to assume that the serialized form will always be in the same place in JSON.

JSON relies on nesting. You can assume that you are in the right place within the "parent" node serialization because it is the one invoking you and once you are invoked, you serialize the nested tree as you know it should be.

> While I like the idea to keep the core of the agent as OTel-agnostic as possible, I don't think that we should handle this right now: ...

I see it exactly the opposite 😄 
Even though it seems it is going this direction, we don't yet know whether using us with OTel is going to be the absolute norm. It seemed to go there with OpenTracing before. Currently we add fields to our core to support niche cases. 
I don't think about other implementations, it's not about abstraction for the purpose of polymorphism, it's abstraction for the purpose of proper responsibility separation - the bridge should know about the core, the core should not know about the bridge.

That said, I am only explaining to emphasize my rationale, it's not something I will insist on so feel free to make the call and leave as is.